File: clients/android/core/src/main/kotlin/com/migo/core/store/ServerEndpoint.kt
package com.migo.core.store

/**
 * The user's chosen server address.
 *
 * Mirrors `packages/sdk/src/server-endpoint.ts` (the canonical JS shape) and
 * `clients/desktop/src/config.rs::ServerEndpoint` (the Rust one), so a self-hoster who
 * types `migo.example.com` into the web form, the desktop settings, or the Android /
 * iOS bootstrap is configuring the same record. The fields are the same six everywhere,
 * the defaults are the same defaults, and the derived URLs come out the same way.
 *
 * The shape is explicit on purpose. A user-typed string that has to be parsed every time
 * it is read is a string that can be parsed two different ways on two different reads;
 * a record the form commits and the bootstrap reads is a record that has one meaning
 * across the whole app, including across a process death and a clean restart.
 *
 * # Why a record and not a URL
 *
 * The string form was `http://host:port/ws`-style, the caller had to build it, and a
 * typo in any byte silently routed the client at a different deployment. A `ServerEndpoint`
 * is host + two ports + two scheme choices, the URL it produces is a function of those
 * fields, and every byte of the URL is something a screen can show before it is asked to
 * be believed.
 *
 * # Why two ports
 *
 * The REST origin and the realtime socket are different listeners, and a reverse-proxy
 * setup (one node, two paths) means they do not have to share a port. The form lets a
 * user leave the gateway port blank and the default is `restPort + 1`, the same one the
 * dev policy uses (`migod` binds REST on 18080 and the gateway on 18081). A user who
 * needs a different split types it.
 *
 * # Why two schemes
 *
 * The realtime transport and the REST control plane are not the same protocol negotiation
 * even when they share a TLS posture. A deployment that fronts both with one TLS terminator
 * still announces WSS and HTTPS, not WSS for both. The split also keeps the loopback policy
 * expressible: a dev server on `localhost` is plain WS and plain HTTP, and a `wss://` to
 * `localhost` would be a user trying to talk to a deployment that is not on this machine.
 */
data class ServerEndpoint(
    /** The lowercased host (no scheme, no port, no path). */
    val host: String,
    /** The REST port. Must be in `[1, 65535]`. */
    val port: Int,
    /** The realtime gateway port. Defaults to [port] + 1. */
    val gatewayPort: Int,
    /** The realtime transport. TCP is the native default; QUIC is a real second option whose wire
     *  path this build still carries over WebSocket. */
    val transport: Transport,
    /** The TLS posture of the realtime transport. */
    val gatewayScheme: GatewayScheme,
    /** The TLS posture of the REST control plane. */
    val restScheme: RestScheme,
) {
    init {
        require(host.isNotBlank()) { "host is required" }
        require(port in 1..65535) { "rest port is out of range (1..65535): $port" }
        require(gatewayPort in 1..65535) { "gateway port is out of range (1..65535): $gatewayPort" }
        // The transport and its scheme must agree: TCP carries TCP/TCP-TLS, a WebSocket carries
        // WS/WSS, QUIC carries QUIC/QUIC-TLS. This is the same pairing the web and desktop forms
        // enforce, so a record that validates here is one every client would accept.
        when (transport) {
            Transport.Tcp -> require(
                gatewayScheme == GatewayScheme.Tcp || gatewayScheme == GatewayScheme.TcpTls,
            ) { "TCP transport requires TCP or TCP-TLS scheme" }
            Transport.WebSocket -> require(
                gatewayScheme == GatewayScheme.Ws || gatewayScheme == GatewayScheme.Wss,
            ) { "WebSocket transport requires WS or WSS scheme" }
            Transport.Quic -> require(
                gatewayScheme == GatewayScheme.Quic || gatewayScheme == GatewayScheme.QuicTls,
            ) { "QUIC transport requires QUIC or QUIC-TLS scheme" }
        }
    }

    /**
     * The REST origin, e.g. `http://localhost:18080`. No trailing slash.
     *
     * This is the form `MigoClientOptions.baseUrl` expects; the gateway URL is derived
     * by the SDK from this one when the caller does not override it.
     */
    fun restBaseUrl(): String = "${restScheme.prefix()}://$host:$port"

    /**
     * The realtime gateway URL, e.g. `ws://localhost:18081/ws`. No trailing slash.
     *
     * The `/ws` path is the path the server exposes; this client does not let the user
     * change it because exposing the path would mean a self-hoster setting it and never
     * knowing which one the server actually answers. The path is the contract.
     */
    fun gatewayUrl(): String = "${gatewayScheme.prefix()}://$host:$gatewayPort/ws"

    companion object {
        /** The REST port the dev policy defaults to, matching `migod` on a fresh install. */
        const val DEFAULT_REST_PORT: Int = 18080

        /**
         * The default endpoint for a host. A loopback gets the dev pair (plain TCP, plain HTTP,
         * gateway on the next port), anything else gets the production pair (WSS over WebSocket,
         * HTTPS, gateway on the same port) — the public deployment serves `/ws` on its single HTTP
         * listener, so WebSocket stays its default until the TCP listener is enabled there.
         *
         * Splitting the rule from the constructor means a settings field that just lost focus can
         * rebuild the pair without re-running the whole endpoint construction.
         */
        fun defaultFor(host: String, port: Int = DEFAULT_REST_PORT): ServerEndpoint {
            // defaultSchemesForHost returns a Pair<GatewayScheme, RestScheme>; destructure
            // it into named locals so the constructor below can refer to each side by name.
            // A Pair has no `gatewayScheme` / `restScheme` fields, so accessing them with
            // `schemes.gatewayScheme` (the previous form) would not compile.
            val (gatewayScheme, restScheme) = defaultSchemesForHost(host)
            return ServerEndpoint(
                host = host.lowercase(),
                port = port,
                gatewayPort = if (restScheme == RestScheme.Https) port else port + 1,
                transport = if (restScheme == RestScheme.Https) Transport.WebSocket else Transport.Tcp,
                gatewayScheme = gatewayScheme,
                restScheme = restScheme,
            )
        }

        /**
         * The dev-policy default: plain TCP on plain HTTP, with the gateway on the next port —
         * the native client's transport pair. `http://localhost:18080` for REST,
         * `tcp://localhost:18081` for the gateway.
         */
        fun loopbackDefault(host: String = "localhost", port: Int = DEFAULT_REST_PORT): ServerEndpoint =
            ServerEndpoint(
                host = host.lowercase(),
                port = port,
                gatewayPort = port + 1,
                transport = Transport.Tcp,
                gatewayScheme = GatewayScheme.Tcp,
                restScheme = RestScheme.Http,
            )

        /**
         * The production default: WSS over HTTPS, with the gateway on the same port as
         * REST. `https://migo.example.com:443` for REST, `wss://migo.example.com:443/ws`
         * for the gateway.
         */
        /** The VPS deployment's single-host endpoint, matching the running migod. */
        fun publicDeploymentDefault(): ServerEndpoint = ServerEndpoint(
            host = "152.53.102.150",
            port = 8080,
            gatewayPort = 8080,
            transport = Transport.WebSocket,
            gatewayScheme = GatewayScheme.Ws,
            restScheme = RestScheme.Http,
        )

        fun internetDefault(host: String, port: Int = 443): ServerEndpoint =
            ServerEndpoint(
                host = host.lowercase(),
                port = port,
                gatewayPort = port,
                transport = Transport.WebSocket,
                gatewayScheme = GatewayScheme.Wss,
                restScheme = RestScheme.Https,
            )

        /**
         * Resolves a REST origin (a `http://host:port` or `https://host:port` string) into
         * a [ServerEndpoint], the same way the legacy sign-in screen did.
         *
         * The bridge the resume path uses: the [SavedSession] still carries a single URL
         * string from before the form was structured, and the bootstrap needs the
         * structured record to build a [MigoClient]. The URL's own scheme is the ground
         * truth for TLS — a session saved against `http://152.53.102.150:8080` must not
         * come back as a TLS endpoint, or the resume handshake dies at the socket and
         * the user is dumped back on the sign-in screen.
         */
        fun fromRestUrl(url: String): ServerEndpoint {
            val trimmed = url.trim()
            if (trimmed.isEmpty()) return loopbackDefault()
            val lower = trimmed.lowercase()
            val (restScheme, authority) = when {
                lower.startsWith("https://") -> RestScheme.Https to trimmed.substring("https://".length)
                lower.startsWith("http://") -> RestScheme.Http to trimmed.substring("http://".length)
                else -> return loopbackDefault()
            }
            val hostAndPort = authority.substringBefore('/').substringBefore('?')
            val (host, port) = when (val colon = hostAndPort.lastIndexOf(':')) {
                -1 -> hostAndPort.lowercase() to if (restScheme == RestScheme.Https) 443 else 80
                else -> hostAndPort.substring(0, colon).lowercase() to
                    (hostAndPort.substring(colon + 1).toIntOrNull() ?: (if (restScheme == RestScheme.Https) 443 else 80))
            }
            // The origin's own scheme decides the posture. A `https://` origin keeps the
            // TLS pair with the gateway on the same port; an `http://` origin keeps the
            // plain pair, with the gateway on the next port only under the dev policy
            // (loopback) — a plain origin on a public host is a single-port deployment
            // like this build's, which serves `/ws` on its HTTP listener.
            return when (restScheme) {
                RestScheme.Https -> internetDefault(host, port)
                RestScheme.Http ->
                    if (isLoopbackHost(host)) {
                        loopbackDefault(host, port)
                    } else {
                        ServerEndpoint(
                            host = host,
                            port = port,
                            gatewayPort = port,
                            transport = Transport.WebSocket,
                            gatewayScheme = GatewayScheme.Ws,
                            restScheme = RestScheme.Http,
                        )
                    }
            }
        }

        /**
         * Picks a default scheme pair for a host. Loopback defaults to the native plain pair
         * (plain TCP, plain HTTP), anything else to the TLS pair (WSS, HTTPS) — the public
         * deployment serves `/ws` on its HTTP listener, so WebSocket stays the non-loopback
         * default until the TCP listener is enabled there.
         */
        fun defaultSchemesForHost(host: String): Pair<GatewayScheme, RestScheme> =
            if (isLoopbackHost(host)) {
                GatewayScheme.Tcp to RestScheme.Http
            } else {
                GatewayScheme.Wss to RestScheme.Https
            }

        /** Hosts the dev policy exempts from the "always TLS" default. */
        fun isLoopbackHost(host: String): Boolean {
            val lowered = host.lowercase()
            return lowered == "localhost" || lowered == "127.0.0.1" || lowered == "::1"
        }
    }
}

/** The realtime transport the user picked. */
enum class Transport {
    /**
     * TCP, the native client's default transport: one connection, one session, binary
     * length-prefixed frames — the mig33v46 heritage. A server advertises it via the
     * `TCP_TRANSPORT` feature bit only when its optional TCP listener is enabled; a node that does
     * not offer the bit falls back to WebSocket, which stays the web client's transport.
     */
    Tcp,

    /**
     * WebSocket, the web client's transport, kept here for development and for the fallback a
     * TCP-less deployment leaves the client on.
     */
    WebSocket,

    /**
     * QUIC, the realtime transport's second option ("opsi kedua"): negotiated via the QUIC feature
     * bit and served by the optional server-side QUIC listener when a deployment enables it. A user
     * can select it and the choice persists as typed. This build has no Kotlin QUIC runtime, so its
     * wire path still connects over WebSocket; a QUIC-capable client honours the persisted endpoint.
     */
    Quic,
}

/** The TLS posture of the realtime gateway. */
enum class GatewayScheme {
    /**
     * Plain TCP. Allowed only for the dev-policy loopback (and a deployment's explicitly plain
     * dev listener); a production TCP listener is fronted by TLS 1.3.
     */
    Tcp,

    /** TCP over TLS 1.3. The native transport's production posture. */
    TcpTls,

    /** Plain WebSocket. Allowed only for loopback hosts; see [ServerEndpoint.defaultFor]. */
    Ws,

    /** WebSocket over TLS. The production form. */
    Wss,

    /**
     * Plain QUIC. The realtime transport's second option, served by the optional server-side QUIC
     * listener when a deployment enables it. Both QUIC postures share the `quic` URL prefix -- the
     * TLS posture rides in ALPN, not the URL.
     */
    Quic,

    /** QUIC over TLS. The second option's production posture; shares the `quic` URL prefix with [Quic]. */
    QuicTls,
    ;

    /** The URL scheme prefix, taking the value's own posture only. */
    fun prefix(): String = when (this) {
        Tcp -> "tcp"
        TcpTls -> "tcps"
        Ws -> "ws"
        Wss -> "wss"
        // Both QUIC variants are spelled `quic` at the URL level; the TLS posture is expressed via
        // ALPN, not the URL. Mirrors web/desktop.
        Quic, QuicTls -> "quic"
    }
}

/** The TLS posture of the REST control plane. */
enum class RestScheme {
    /** Plain HTTP. Allowed only for loopback hosts. */
    Http,

    /** HTTPS. The production form. */
    Https,
    ;

    /** The URL scheme prefix, taking the value's own posture only. */
    fun prefix(): String = when (this) {
        Http -> "http"
        Https -> "https"
    }
}

File: packages/sdk/src/server-endpoint.ts
/**
 * The user's chosen server address.
 *
 * The bootstrap surface used to be a single string the caller had to build — `https://node.example/ws`,
 * say — and a typo in any byte silently routed the client at a different deployment. A user who
 * self-hosts wants a chance to type the address, see what they typed, and not have the SDK then turn
 * that into an obscure URL on their behalf. This is the small, explicit shape that lets them do
 * that: every field is something they can change in the form, every derived URL comes from those
 * fields by a rule the file documents, and a one-line shorthand is parsed without ever becoming a
 * second way to express the same thing.
 *
 * The transport and scheme enums are deliberately split: a WebSocket on `WS` is the plain-TCP dev
 * affordance, a WebSocket on `WSS` is the production form, and QUIC is the second realtime transport
 * option. A server advertises QUIC via the `QUIC` feature bit only when its optional QUIC listener
 * is enabled (a QUIC-capable host offers `FEATURE.QUIC` itself through `hello.features`). This SDK
 * validates and persists the choice; the data path it opens is still WebSocket.
 */

/** The transport the realtime socket speaks. */
export type Transport = 'WebSocket' | 'Quic';

/** The TLS posture of the WebSocket transport. */
export type WsScheme = 'Ws' | 'Wss';

/** The TLS posture of the QUIC transport. */
export type QuicScheme = 'Quic' | 'QuicTls';

/** The TLS posture for the REST surface. */
export type RestScheme = 'Http' | 'Https';

/** The scheme paired with the chosen transport. */
export type Scheme = WsScheme | QuicScheme;

/** Hosts that are exempt from the "always use TLS" default. */
const LOOPBACK_HOSTS: ReadonlySet<string> = new Set(['localhost', '127.0.0.1', '::1']);

/**
 * A parsed host, lowercased and with any inline `:port` already split off into {@link ServerEndpoint.port}.
 */
export type Host = string;

/**
 * The user-configured server: one host, one REST port, one gateway port, and the pair of schemes that
 * tell the client whether each side is plain or encrypted.
 *
 * The two ports are split on purpose. They default together (gateway = rest + 1) but the form lets a
 * user who has to, because of a reverse-proxy setup, point the REST origin and the realtime socket
 * at different listeners without either side being magic. The transport enum is the only one that has
 * to grow when a new realtime path lands; the schemes are already expressed at the level the form
 * and the protocol both speak.
 */
export interface ServerEndpoint {
  /** The lowercased host (no scheme, no port, no path). */
  host: Host;
  /** The REST port. Must be in `[1, 65535]`. */
  port: number;
  /** The gateway port. Defaults to `port + 1`; the form exposes an override. */
  gatewayPort: number;
  /** The realtime transport. */
  transport: Transport;
  /** The TLS posture of the realtime transport. */
  scheme: Scheme;
  /** The TLS posture of REST, paired with the realtime choice when sensible. */
  restScheme: RestScheme;
}

/** What a {@link parseServerEndpoint} call can return on a bad input. */
export class ServerEndpointError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ServerEndpointError';
  }
}

/**
 * Builds a {@link ServerEndpoint} for a host that is a development loopback.
 *
 * The default schemes are the only one the dev policy allows: a plain WebSocket and a plain HTTP on
 * the same port, with the gateway on the next one up. A `wss://` request to a loopback would be a
 * user trying to talk to a TLS-only deployment that has nothing to do with their local node, and
 * the form does not need to enable it.
 */
export function defaultLoopbackServerEndpoint(host: string, port = 18080): ServerEndpoint {
  return {
    host: host.toLowerCase(),
    port,
    gatewayPort: port + 1,
    transport: 'WebSocket',
    scheme: 'Ws',
    restScheme: 'Http',
  };
}

/**
 * Builds a {@link ServerEndpoint} for a production-style address.
 *
 * A non-loopback host forces the TLS postures. The REST origin speaks HTTPS and the gateway speaks
 * WSS — the only configuration the audit allows once a deployment is reachable from outside this
 * machine, and the form's default when the user types a real domain.
 */
export function defaultInternetServerEndpoint(host: string, port = 443): ServerEndpoint {
  return {
    host: host.toLowerCase(),
    port,
    gatewayPort: port,
    transport: 'WebSocket',
    scheme: 'Wss',
    restScheme: 'Https',
  };
}

/**
 * True when the host is one the local-dev policy applies to: a plain WebSocket is allowed there,
 * a `https://` request would point at a TLS-only deployment that has nothing to do with this machine.
 */
export function isLoopbackHost(host: string): boolean {
  return LOOPBACK_HOSTS.has(host.toLowerCase());
}

/**
 * Picks a default scheme pair for a host. Loopback defaults to plain (dev), anything else to TLS.
 *
 * Splitting the rule from the constructor means a form field that just lost focus can rebuild its
 * pair without re-running the whole endpoint construction, and a unit test can pin the rule without
 * standing up a full endpoint.
 */
export function defaultSchemesForHost(host: string): { scheme: WsScheme; restScheme: RestScheme } {
  if (isLoopbackHost(host)) {
    return { scheme: 'Ws', restScheme: 'Http' };
  }
  return { scheme: 'Wss', restScheme: 'Https' };
}

/**
 * Parses the two shorthand shapes a user is likely to type, and the split form for everything else.
 *
 * Three inputs are recognised:
 *
 *   - `host` — bare, e.g. `migo.example.com`. The port defaults to {@link DEFAULT_REST_PORT}.
 *   - `host:port` — a single colon and a numeric port, e.g. `migo.example.com:8443`. Anything else
 *     (no colon, no number) is rejected.
 *   - Two arguments, the host and the port. The form uses this when the user has typed the two
 *     fields separately.
 *
 * The IPv6 form (`[::1]:18080`) is intentionally not supported yet; the brief is dev/local, the
 * production path uses hostnames, and adding a parser for a case the form has no field for would
 * mean untested code in the hot path of a register screen.
 */
export function parseHost(
  input: string,
  portFallback = DEFAULT_REST_PORT,
): { host: string; port: number } {
  const trimmed = input.trim();
  if (trimmed === '') {
    throw new ServerEndpointError('host is required');
  }
  const colon = trimmed.indexOf(':');
  if (colon < 0) {
    return { host: trimmed.toLowerCase(), port: portFallback };
  }
  // Multiple colons that are not the IPv6 bracket form are not a host:port the form can take.
  if (trimmed.indexOf(':', colon + 1) >= 0) {
    throw new ServerEndpointError(`host cannot contain more than one colon: ${trimmed}`);
  }
  const host = trimmed.slice(0, colon).toLowerCase();
  const portText = trimmed.slice(colon + 1);
  if (host === '') {
    throw new ServerEndpointError(`host is empty: ${trimmed}`);
  }
  const port = Number.parseInt(portText, 10);
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new ServerEndpointError(`port is out of range (1..65535): ${portText}`);
  }
  if (portText !== port.toString()) {
    // Reject `8080abc` and friends — `parseInt` would accept them.
    throw new ServerEndpointError(`port is not a whole number: ${portText}`);
  }
  return { host, port };
}

/** The REST port used when none is supplied. Matches the dev policy default. */
export const DEFAULT_REST_PORT = 18080;
/** The gateway port used when none is supplied. Defaults to `restPort + 1`. */
export const DEFAULT_GATEWAY_PORT_OFFSET = 1;

/**
 * Validates the ranges on the numeric fields. Split out so the constructor and the parser share it.
 */
export function validatePorts(port: number, gatewayPort: number): void {
  if (!Number.isInteger(port) || port < 1 || port > 65535) {
    throw new ServerEndpointError(`rest port is out of range (1..65535): ${port}`);
  }
  if (!Number.isInteger(gatewayPort) || gatewayPort < 1 || gatewayPort > 65535) {
    throw new ServerEndpointError(`gateway port is out of range (1..65535): ${gatewayPort}`);
  }
}

/**
 * Result type for {@link validate} on a {@link ServerEndpoint}.
 *
 * A `Result<ServerEndpoint, Error>` is the only public face; the constructors throw on a bad shape
 * (a caller that hands a malformed endpoint in gets a clear error), and the validators return the
 * same shape so a form can show the same message either way.
 */
export interface ServerEndpointValidation {
  /** `null` when the endpoint is well-formed. */
  error: string | null;
}

const VALID_REST_SCHEMES: ReadonlySet<RestScheme> = new Set<RestScheme>(['Http', 'Https']);
const VALID_WS_SCHEMES: ReadonlySet<WsScheme> = new Set<WsScheme>(['Ws', 'Wss']);
const VALID_QUIC_SCHEMES: ReadonlySet<QuicScheme> = new Set<QuicScheme>(['Quic', 'QuicTls']);

/**
 * Checks every field of an endpoint and returns the first failure, or `null` when the shape is sound.
 *
 * The four rules are the same ones a user-visible form rejects on:
 *
 *   1. `host` is not empty.
 *   2. `port` and `gatewayPort` are integers in `[1, 65535]`.
 *   3. The `(transport, scheme)` pair is consistent: a WebSocket transport takes a `Ws` or `Wss`
 *      scheme and never a `Quic*` scheme, and the QUIC transport takes a `Quic*` scheme and
 *      never a `Ws*` scheme.
 *   4. `restScheme` is `Http` or `Https`.
 *
 * The rules are deliberately written as four checks rather than a long boolean expression: each one
 * names the field it is about, so the message a form shows can be the message a test pins.
 */
export function validateServerEndpoint(endpoint: ServerEndpoint): ServerEndpointValidation {
  if (typeof endpoint.host !== 'string' || endpoint.host.trim() === '') {
    return { error: 'host is required' };
  }
  if (!Number.isInteger(endpoint.port) || endpoint.port < 1 || endpoint.port > 65535) {
    return { error: `rest port is out of range (1..65535): ${String(endpoint.port)}` };
  }
  if (
    !Number.isInteger(endpoint.gatewayPort) ||
    endpoint.gatewayPort < 1 ||
    endpoint.gatewayPort > 65535
  ) {
    return { error: `gateway port is out of range (1..65535): ${String(endpoint.gatewayPort)}` };
  }
  if (endpoint.transport === 'WebSocket') {
    if (!VALID_WS_SCHEMES.has(endpoint.scheme as WsScheme)) {
      return { error: 'WebSocket transport requires WS or WSS scheme' };
    }
  } else {
    if (!VALID_QUIC_SCHEMES.has(endpoint.scheme as QuicScheme)) {
      return { error: 'QUIC transport requires QUIC or QUIC-TLS scheme' };
    }
  }
  if (!VALID_REST_SCHEMES.has(endpoint.restScheme)) {
    return { error: 'REST scheme must be HTTP or HTTPS' };
  }
  return { error: null };
}

/**
 * Throws {@link ServerEndpointError} when {@link validateServerEndpoint} returned an error.
 *
 * Split from the validators that return a result so a constructor can call the throwing form and
 * have a single error path, while the form calls the result form and renders the message.
 */
export function assertValidServerEndpoint(endpoint: ServerEndpoint): void {
  const result = validateServerEndpoint(endpoint);
  if (result.error !== null) {
    throw new ServerEndpointError(result.error);
  }
}

/**
 * Returns the gateway scheme prefix, taking the transport into account.
 *
 * Split from the URL builder because the form's display logic wants to show the scheme alone (a
 * select), and a test pins the rule with a plain `assert.equal` rather than a full URL parse.
 */
export function gatewaySchemePrefix(endpoint: ServerEndpoint): string {
  if (endpoint.transport === 'Quic') {
    // Both QUIC schemes map to the single "quic" URL prefix: the TLS posture rides in the REST
    // scheme, not the gateway scheme. This matches the Rust and Kotlin clients.
    return 'quic';
  }
  return endpoint.scheme === 'Wss' ? 'wss' : 'ws';
}

/** The REST scheme prefix for the URL, taking the `restScheme` field on its own. */
export function restSchemePrefix(endpoint: ServerEndpoint): string {
  return endpoint.restScheme === 'Https' ? 'https' : 'http';
}

/**
 * The REST origin, e.g. `http://localhost:18080`. No trailing slash.
 *
 * Used by {@link BootstrapClient} as its base URL; the trailing-slash normalisation the bootstrap
 * client applies internally handles the rest.
 */
export function restBaseUrl(endpoint: ServerEndpoint): string {
  return `${restSchemePrefix(endpoint)}://${endpoint.host}:${endpoint.port}`;
}

/**
 * The gateway WebSocket URL, e.g. `ws://localhost:18081/ws`. No trailing slash.
 *
 * The `/ws` path is the path the server exposes; the SDK does not let callers change it because
 * exposing the path would mean a self-hoster setting it and never knowing which one the server
 * actually answers. The path is the contract.
 */
export function gatewayUrl(endpoint: ServerEndpoint): string {
  return `${gatewaySchemePrefix(endpoint)}://${endpoint.host}:${endpoint.gatewayPort}/ws`;
}

/**
 * Resolves a server URL string (the env or the legacy form) into a {@link ServerEndpoint}.
 *
 * `MIGO_PUBLIC_URL` and the dev default both arrive as a REST origin, e.g. `http://localhost:18080`
 * or `http://migo.example`. The form gives the SDK structured fields today; the env gives it a
 * string from before the form existed. This is the bridge: take the string, derive the
 * endpoint, and let the rest of the SDK forget it was ever a string.
 *
 * The URL's own scheme is the ground truth for both postures — the same rule the Android and
 * desktop clients apply:
 *
 *   - `https://…` is the production pair (`Wss`/`Https`) with the gateway on the same port;
 *   - `http://…` on a loopback host is the dev pair (`Ws`/`Http`) with the gateway on the next
 *     port (the split-port dev policy);
 *   - `http://…` on any other host is a single-port plain deployment (this repository's own
 *     VPS shape: `migod` serves `/ws` on its HTTP listener), so the gateway rides the same
 *     port, plain.
 */
export function serverEndpointFromUrl(url: string): ServerEndpoint {
  const trimmed = url.trim();
  if (trimmed === '') {
    return defaultLoopbackServerEndpoint('localhost');
  }
  let parsed: URL;
  try {
    parsed = new URL(trimmed);
  } catch {
    throw new ServerEndpointError(`not a valid URL: ${url}`);
  }
  const scheme = parsed.protocol === 'https:' ? 'Https' : 'Http';
  const restPort =
    parsed.port === '' ? (scheme === 'Https' ? 443 : 80) : Number.parseInt(parsed.port, 10);
  const host = parsed.hostname.toLowerCase();
  // The URL's own scheme decides both postures (see the doc above). Only a plain origin on a
  // loopback host keeps the dev split — gateway on the next port; everything else serves `/ws`
  // on the port the URL names.
  const gatewayScheme: WsScheme = scheme === 'Https' ? 'Wss' : 'Ws';
  const gatewayPort =
    scheme === 'Https' || !isLoopbackHost(host) ? restPort : restPort + DEFAULT_GATEWAY_PORT_OFFSET;
  return {
    host,
    port: restPort,
    gatewayPort,
    transport: 'WebSocket',
    scheme: gatewayScheme,
    restScheme: scheme,
  };
}
